crates/turborepo-telemetry/src/lib.rs

```diff
@@ -0,0 +1,424 @@
+                    biased;
+                    event = self.rx.recv() => {
+                        if let Some(event) = event {
+                            self.buffer.push(event);
+                        } else {
+                            // There are no senders left so we can shut down
+                            break;
+                        }
+                        if self.buffer.len() == BUFFER_THRESHOLD {
+                            self.flush_events();
+                            timeout = tokio::time::sleep(NO_TIMEOUT);
+                        } else {
+                            timeout = tokio::time::sleep(EVENT_TIMEOUT);
+                        }
+                    }
+                    _ = timeout => {
+                        self.flush_events();
+                        timeout = tokio::time::sleep(NO_TIMEOUT);
+                    }
+                    _ = self.exit_ch.closed() => {
+                        break;
+                    }
+                }
+            }
+            self.flush_events();
+            while let Some(result) = self.senders.next().await {
+                if let Err(err) = result {
+                    debug!("failed to send telemetry event. error: {}", err)
+                }
+            }
+        })
+    }
+
+    pub fn flush_events(&mut self) {
+        if !self.buffer.is_empty() {
+            let events = std::mem::take(&mut self.buffer);
+            debug!(
+                "Starting telemetry event queue flush (num_events={:?})",
+                events.len()
+            );
+            let handle = self.send_events(events);
+            self.senders.push(handle);
+            debug!("Done telemetry event queue flush");
+        }
+    }
+
+    fn send_events(&self, events: Vec<TelemetryEvent>) -> JoinHandle<()> {
+        if !self.enabled {
+            // do nothing
+            return tokio::spawn(async move {});
+        }
+
+        if config::is_debug() {
+            for event in &events {
+                let pretty_event = serde_json::to_string_pretty(&event)
+                    .unwrap_or("Error serializing event".to_string());
+                println!("");
+                println!(
+                    "\n{}\n{}\n",
+                    self.ui.apply(BOLD.apply_to("[telemetry event]")),
+                    self.ui.apply(GREY.apply_to(pretty_event))
+                );
+            }
+        }
+
+        let client = self.client.clone();
+        let session_id = self.session_id.clone();
+        let telemetry_id = self.telemetry_id.clone();
+        tokio::spawn(async move {
+            if let Ok(Err(err)) = tokio::time::timeout(
+                REQUEST_TIMEOUT,
+                client.record_telemetry(events, telemetry_id.as_str(), session_id.as_str()),
+            )
+            .await
+            {
+                debug!("failed to record cache usage telemetry. error: {}", err)
+            }
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I still need to update these tests

Reply:
Updated for now, but may change this in the future, `OnceCell` is very annoying to test.